src/main/java/com/google/sps/servlets/LoginServlet.java

```diff
@@ -16,74 +16,69 @@
 import javax.servlet.http.HttpServletResponse;
 import com.google.gson.Gson;
 import com.google.sps.data.User;
+import java.util.Map;
+import java.util.HashMap;
 
 @WebServlet("/login")
 public class LoginServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-
     UserService userService = UserServiceFactory.getUserService();
 
     String urlToRedirect = "/";
     String userId = "null";
     String logoutUrl = "null";
     String loginUrl = "null";
     String userEmail = "null";
-
-    // Boolean to update User's main page 
     Boolean showNewTab = false;
 
+    User user;
     if (userService.isUserLoggedIn()) {
-      userEmail = userService.getCurrentUser().getEmail();
       userId = userService.getCurrentUser().getUserId();
+      userEmail = userService.getCurrentUser().getEmail();
       logoutUrl = userService.createLogoutURL(urlToRedirect);
       showNewTab = true;
     } else {
       loginUrl = userService.createLoginURL(urlToRedirect);
     }
 
-    User user = new User(userId, loginUrl, logoutUrl, userEmail, showNewTab);
-    
-    if (userId != "null") {
-        checkUserInDatastore(user);
+    user = new User(userId, userEmail);
+
+    if (userId != "null" && !isUserInDatastore(user)) {
+      storeUserToDatastore(user);
     }
 
+    Map<String, Object> jsonInfo = new HashMap<>();
+    jsonInfo.put("showTabStatus", showNewTab);
+    jsonInfo.put("logoutUrl", logoutUrl);
+    jsonInfo.put("loginUrl", loginUrl);
+
     response.setContentType("application/json;");
-    response.getWriter().println(new Gson().toJson(user));
+    response.getWriter().println(new Gson().toJson(jsonInfo));
   }
 
-  public void checkUserInDatastore(User user) {
-    // ** NEED TO WRITE TEST FOR BEHAVIORAL TESTING
-
-    Key userKey = KeyFactory.createKey("User", user.getUserId());
+  public boolean isUserInDatastore(User user) {
     DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
-    Entity userEntity;
+    Key userKey = KeyFactory.createKey("User", user.getUserId());
 
-    // Check if user is in datastore
     try {
-      userEntity = datastore.get(userKey);
+      datastore.get(userKey);
       user.setUserKey(KeyFactory.keyToString(userKey));
```

Comment:
You don't have to do this in this PR, but I would suggest restructuring this method. Currently, the name of the method suggests that it's a pure function — it looks up whether the user is in Datastore. But it actually has side effects — it alters user itself.

I think a better way to write this function to eliminate side effects is to return the Key if the user exists in Datastore — otherwise, return null. Then the calling code can use this to determine whether the user already exists, and the side effect can be done there too.

Reply:
Thank you! I'll fix this in a different branch. I'll maintain this concept in the future